boa_engine/src/property/nonmaxu32.rs

```diff
@@ -0,0 +1,35 @@
+/// An integer that is not `u32::MAX`.
+#[derive(PartialEq, Debug, Clone, Copy, Eq, Hash)]
+pub struct NonMaxU32 {
+    inner: u32,
+}
+
+impl NonMaxU32 {
+    /// Creates a non-max `u32`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the given value is not `u32::MAX`.
+    #[must_use]
+    pub const fn new_unchecked(inner: u32) -> Self {
+        debug_assert!(inner != u32::MAX);
+
+        Self { inner }
+    }
```

Comment:
I think this should be an unsafe function. My reasoning is that, in the future, we could implement an optimization that assumes that the invariant holds. However, if leave this function as safe, we'd have to make a breaking change to add `unsafe` at that moment. So, in preparation to that I'd mark it as unsafe.

Reply:
I would argue that if we need to change this to be unsafe later, we just do the breaking change. In my opinion it is more important to have a clean API that does what is expected, than to avoid breaking changes.

Reply:
It's not only a matter of doing the breaking change though. If, for example, someone tried to do an unsafe optimization assuming the invariant holds, and the invariant didn't hold, it would mean that a safe function would be the cause of UB.

Reply:
E.g. [`String::from_utf8_unchecked`](https://doc.rust-lang.org/std/string/struct.String.html#safety-1) with it's safety comment.

Reply:
Well but currently we do not have any optimizations based on the invariant, right? If we add some, we would have to mark the function as unsafe in that same change. In contrast I think there are unsafe optimizations in `String` that debend on the bytes being UTF-8.

Reply:
> Well but currently we do not have any optimizations based on the invariant, right?

Note that the safety comment doesn't specify that there are uses of that method that unsafely assume the string is utf8.
> If this constraint is violated, it may cause memory unsafety issues with future users (sic) of the String, ...
Emphasis on "... with future uses of the String"

> If we add some, we would have to mark the function as unsafe in that same change.

I don't think it's ideal to rely on the fact that all of us have to remember to change the function to `unsafe` in the same PR.

Reply:
I think the `String` documentation is not talking about "future" in the sense of functionality that may or may be not added in the future but rather the future of the created `String` itself. I think the next sentence makes it a little more clear: `[...] as the rest of the standard library assumes that Strings are valid UTF-8.`

> I don't think it's ideal to rely on the fact that all of us have to remember to change the function to unsafe in the same PR.

I very much disagree with this. When we add unsafe optimizations, we have to check that there is no way the constraints that the optimizations depend on can be violated outside of unsafe functions themselves. In this case, we would have to make sure that there is no way a `NonMaxU32` can be constructed without a `u32::MAX` value, expect from an unsafe function that documents this constraint.

Nevermind all of that, let's not get too hung up on this. If you think the function should be unsafe, I will add it back and remove the debug_assert.